exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/agg/TestHashAggrSpill.java

```diff
@@ -0,0 +1,141 @@
+        //profile.print();
+        List<ProfileParser.OperatorProfile> ops = profile.getOpsOfType(UserBitShared.CoreOperatorType.HASH_AGGREGATE_VALUE);
+
+        assertTrue( ! ops.isEmpty() );
+        // check for the first op only
+        ProfileParser.OperatorProfile hag0 = ops.get(0);
+        long opCycle = hag0.getMetric(HashAggTemplate.Metric.SPILL_CYCLE.ordinal());
+        assertEquals(spillCycle, opCycle);
+        long op_spilled_partitions = hag0.getMetric(HashAggTemplate.Metric.SPILLED_PARTITIONS.ordinal());
+        assertEquals(spilledPartitions, op_spilled_partitions);
+        /* assertEquals(3, ops.size());
+        for ( int i = 0; i < ops.size(); i++ ) {
+            ProfileParser.OperatorProfile hag = ops.get(i);
+            long cycle = hag.getMetric(HashAggTemplate.Metric.SPILL_CYCLE.ordinal());
+            long num_partitions = hag.getMetric(HashAggTemplate.Metric.NUM_PARTITIONS.ordinal());
+            long spilled_partitions = hag.getMetric(HashAggTemplate.Metric.SPILLED_PARTITIONS.ordinal());
+            long mb_spilled = hag.getMetric(HashAggTemplate.Metric.SPILL_MB.ordinal());
+            System.out.println(String.format("(%d) Spill cycle: %d, num partitions: %d, spilled partitions: %d, MB spilled: %d", i,cycle, num_partitions, spilled_partitions,
+                mb_spilled));
+        } */
+    }
+
+    /**
+     * Test "normal" spilling: Only 2 partitions (out of 4) would require spilling
+     * ("normal spill" means spill-cycle = 1 )
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testHashAggrSpill() throws Exception {
+        LogFixture.LogFixtureBuilder logBuilder = LogFixture.builder()
+            .toConsole()
+            .logger("org.apache.drill.exec.physical.impl.aggregate", Level.WARN)
+            ;
+
+        FixtureBuilder builder = ClusterFixture.builder()
+            .configProperty(ExecConstants.HASHAGG_MAX_MEMORY_KEY,62_000_000)
+            .configProperty(ExecConstants.HASHAGG_NUM_PARTITIONS_KEY,16)
+            .configProperty(ExecConstants.HASHAGG_MIN_BATCHES_PER_PARTITION_KEY,3)
+            .sessionOption(PlannerSettings.FORCE_2PHASE_AGGR_KEY,true)
+            // .sessionOption(PlannerSettings.EXCHANGE.getOptionName(), true)
+            .maxParallelization(2)
+            .saveProfiles()
+            //.keepLocalFiles()
+            ;
+        try (LogFixture logs = logBuilder.build();
+             ClusterFixture cluster = builder.build();
+             ClientFixture client = cluster.clientFixture()) {
+            String sql = "SELECT empid_s17, dept_i, branch_i, AVG(salary_i) FROM `mock`.`employee_1200K` GROUP BY empid_s17, dept_i, branch_i";
+            runAndDump(client, sql, 1_200_000, 1, 1);
+        }
+    }
+
+    /**
+     *  Test "secondary" spilling -- Some of the spilled partitions cause more spilling as they are read back
+     *  (Hence spill-cycle = 2 )
+     *
+     * @throws Exception
+     */
+    @Test
```

Comment:
Question: are two tests sufficient to test all the code paths in the modified code? Might we need test for specific cases? Will provide detailed suggestions as I review the actual spill code.